src/ft/io/Reader.zig

```diff
@@ -0,0 +1,220 @@
+const ft = @import("../ft.zig");
+const mem = ft.mem;
+
+context: *const anyopaque,
+
+readFn: *const fn (*const anyopaque, []u8) anyerror!usize,
+
+const Self = @This();
+
+pub fn discard(self: Self) anyerror!u64 {
+    var buffer: [100]u8 = undefined;
+    var ret: u64 = 0;
+    while (true) {
+        const tmp = try self.read(buffer[0..]);
+        ret += tmp;
+        if (tmp == 0)
+            break;
+    }
+    return ret;
+}
+
+pub fn isBytes(self: Self, slice: []const u8) anyerror!bool {
+    var ret = true;
+    for (slice) |c| {
+        ret = (ret and (try self.readByte()) == c);
+    }
+    return ret;
+}
+
+pub fn read(self: Self, buffer: []u8) anyerror!usize {
+    return self.readFn(self.context, buffer);
+}
+
+pub fn readAll(self: Self, buffer: []u8) anyerror!usize {
+    return self.readAtLeast(buffer, buffer.len);
+}
+
+pub fn readAllAlloc(self: Self, allocator: mem.Allocator, max_size: usize) anyerror![]u8 {
+    var array = ft.ArrayList(u8).init(allocator);
+    defer array.deinit();
+    try self.readAllArrayList(&array, max_size);
+    array.shrinkAndFree(array.slice.len);
+    return array.toOwnedSlice();
+}
+
+pub fn readAllArrayList(self: Self, array_list: *ft.ArrayList(u8), max_append_size: usize) anyerror!void {
+    return self.readAllArrayListAligned(1, array_list, max_append_size); // todo align
+}
+
+pub fn readAllArrayListAligned(
+    self: Self,
+    comptime alignment: ?u29,
+    array_list: *ft.ArrayListAligned(u8, alignment),
+    max_append_size: usize,
+) anyerror!void {
+    var buffer: [128]u8 = undefined;
+    var remaining = max_append_size;
+
+    while (true) {
+        const size = try self.read(buffer[0..]);
+        if (size == 0) {
+            break;
+        }
+        try array_list.appendSlice(buffer[0..@min(size, remaining)]);
+        if (remaining < size) {
+            return error.StreamTooLong;
+        }
+        remaining -= size;
+    }
+}
+
+pub fn readAtLeast(self: Self, buffer: []u8, len: usize) anyerror!usize {
+    ft.debug.assert(len <= buffer.len);
+    var ret: usize = 0;
+    while (ret < len) {
+        const tmp = try self.read(buffer[ret..]);
+        ret += tmp;
+        if (tmp == 0) {
+            break;
+        }
+    }
+    return ret;
+}
+
+// pub fn readBoundedBytes(self: Self, comptime num_bytes: usize) anyerror!std.BoundedArray(u8, num_bytes)
+
+pub fn readByte(self: Self) anyerror!u8 {
+    var buffer: [1]u8 = undefined;
+    try self.readNoEof(buffer[0..]);
+    return buffer[0];
+}
+
+pub fn readByteSigned(self: Self) anyerror!i8 {
+    return @bitCast(self.readByte());
```

Comment:
```suggestion
    return @bitCast(try self.readByte());
```
I'm not 100% sure, but i feel it makes more sens to `try` to readByte to properly return the error if an error occurs during readByte call, doesn't it?